Task2 module 2

robinho1989 opened the pull request:
Hi Boss :) I know, that this task is not completely finished, but I had some difficulties. I could not solve them.
1. Display of playing animation does not work correctly. I could not find a reason why.
2. Something is wrong with feeding and sleeping states, when the tamgotchi is 'dead' in particular.
I am sure you will find more mistakes, so I am waiting for response :)

robinho1989 committed:
display of image created

robinho1989 committed:
tamagotchi states created

Sumick commented on js/modules/tamagotchi.js:
```diff
@@ -4,7 +4,34 @@ export default class Tamagotchi {
 		this.energy = { value: 10, importance: 2, element: null };
 		this.fun = { value: 10, importance: 4, element: null };
+
+		this.initial = {
```
**praise:** Nice state split! It looks clear, and readable.

Sumick replied:
**nitpick:** It should be a happy state - not initial.

Sumick commented on js/modules/tamagotchi.js:
```diff
@@ -4,7 +4,34 @@ export default class Tamagotchi {
 		this.fun = { value: 10, importance: 4, element: null };
+
+		this.initial = {
+			state: 'Happy',
```
**nitpick:** I would close these states in one `this.states` object where you keep could them

Sumick commented on js/modules/tamagotchi.js:
```diff
@@ -27,6 +54,58 @@ export default class Tamagotchi {
+	displayState = (imageSelector) => {
+		const image = document.querySelector(imageSelector);
+		if (
+			this.energy.value &&
```
**nitpick:** You don't have any single which your Tamagotchi currently have. Doing it this way will probably overwrite your actions, and it will make impossible to control. Here is good example - how (based on your current implementation) would you check what state your Tamagotchi currently have? You would have to check CSS class - but it's out of JS logic, which is a very bad idea. You don't have any control over it.

Try to implement something like: `this.activeState`, and then you could just do simple `switch` which could look like eg.:

```
switch(this.currentState) {
        case this.states.sleepy.stateName:
           image.src = this.states.sleepy.path;
	   this.displayParagraph(this.states.sleepy.stateName);
           break;
}
           ...
```

Sumick commented on js/modules/tamagotchi.js:
```diff
@@ -70,17 +152,65 @@ export default class Tamagotchi {
+	feedingAction = (elementSelector) => {
+		const feedingButton = document.querySelector(elementSelector);
+		const image = document.querySelector('.tamagotchiImage');
+		feedingButton.addEventListener('mousedown', () => {
```
**nitpick:** Logic inside could be as simple as eg: 
```
this.activeState=this.states.eating.stateName
```

Sumick reviewed (changes requested):
It does not work because you are drowning in 'ifology', and it is really hard to predict your current logic. It is because you are missing one, very important thing - your single, active state that your Tamgo currently have. If you refactor this code using my advise from comments below, it should start working - or at least it should be clear enough to understand why it does not work :-)

To sum up:

- [x] Implement `this.activeState`, and refactor your code according to that
- [x] Use `switch` statement to display Tamago UI (based on `this.activeState`)
- [x] Pack your states into `this.states` object, so they are together

robinho1989 committed:
activeState and switch  added

robinho1989 committed:
abilities created

robinho1989 committed:
states implemented

robinho1989 committed:
animations added

robinho1989 commented:
Hi. I managed to fix some issues. I created new state to update Tamago states. Thanks to it all states work. But I found another issue with playing animation. I used one animation for feeding and sleeping and it works fine but it is a problem with playing state animation. I had to create another animation but it still does not work properly.

robinho1989 committed:
first attempt of fixin the playing animation

robinho1989 commented:
I think that the problem is caused by size of sprite sheet. First two animations have the same size, but when I want to import files for last animation they have different size. I tried to cut them in editor and create a new sprite sheet but effect which I achieved is poor :(

Sumick reviewed (changes requested):
I've checked it, and it's not a problem with sprite size, but with `overflow: hidden` that cuts your sprite. Solution here would be extending your image container classes. And then just place your sprite accordingly using your keyframes. 

PS. One additional thing regarding your initial state. Tamago seems to be disappeared for initial moment ([video](https://jmp.sh/6wO1NEA4))

<img width="949" alt="Zrzut ekranu 2023-04-3 o 14 36 55" src="https://user-images.githubusercontent.com/22893188/229511443-4bfd84aa-0b21-4391-b75c-b00039e22940.png">


<img width="452" alt="Zrzut ekranu 2023-04-3 o 14 33 43" src="https://user-images.githubusercontent.com/22893188/229510697-411ce1b6-684d-404e-95ef-9f5d6f51dc0e.png">

robinho1989 committed:
initial image fixed

robinho1989 commented:
Hi. I fixed an initial image display, aniamtions for feeding and sleeping. As you mentioned I added new classes for each state for container where an image is placed. Playing animation works but I got lost with positioning. Unfortunately I don't know how to place it to display last animation properly.

Sumick reviewed (approved):
Code looks good, and task scope is properly covered. Im approving, so you can push this project further. We can get back to this animation positioning issue at the very end, as a part of project polishing.
